scripts/core/editor3/helpers/insertAtomicBlockWithoutEmptyLinesAroundIt.js

```diff
@@ -0,0 +1,95 @@
+import {
+    BlockMapBuilder,
+    CharacterMetadata,
+    ContentBlock,
+    Modifier,
+    EditorState,
+    genKey
+} from 'draft-js';
+
+const Immutable = require('immutable');
+
+var List = Immutable.List,
+    Repeat = Immutable.Repeat;
+
+function getInsertionTarget(contentState, selectionState) {
+    var afterRemoval = Modifier.removeRange(contentState, selectionState, 'backward');
+    var targetSelection = afterRemoval.getSelectionAfter();
+    var afterSplit = Modifier.splitBlock(afterRemoval, targetSelection);
+
+    // return unsplit content if after the splitting first line is empty
+    if (afterSplit.getBlockForKey(afterSplit.getSelectionBefore().getAnchorKey()).getText().length < 1) {
+        return {
+            contentState: afterRemoval,
+            selectionState: targetSelection
+        };
+    }
+
+    var insertionTarget = afterSplit.getSelectionAfter();
+
+    return {
+        contentState: afterSplit,
+        selectionState: insertionTarget
+    };
+}
+
+/*
+    insertAtomicBlockWithoutEmptyLinesAroundIt: function(
+        editorState: EditorState,
+        entityKey: string,
+        character: string
+    ): EditorState
+*/
+
+function insertAtomicBlockWithoutEmptyLinesAroundIt(editorState, entityKey, character) {
+    var selectionState = editorState.getSelection();
+    var target = getInsertionTarget(editorState.getCurrentContent(), selectionState);
+    var asAtomicBlock = Modifier.setBlockType(target.contentState, target.selectionState, 'atomic');
+    var charData = CharacterMetadata.create({entity: entityKey});
+    var fragmentArray = [];
+
+    if (asAtomicBlock.getFirstBlock().getKey() === target.selectionState.getAnchorKey()) {
+        fragmentArray.push(new ContentBlock({
```

Comment:
we keep annotations/comments data on first block atm, would this still work if there is new block added?

Reply:
I've tried to break it by inserting a table in various positions when having active highlights. Seems to work fine.
![highlights](https://user-images.githubusercontent.com/2076953/35607791-b1a7df46-065e-11e8-8284-a81eee8deedc.gif)